sdk/identity/Azure.Identity/src/ImdsManagedIdentityProbeSource.cs

```diff
@@ -127,12 +130,29 @@ public async override ValueTask<AccessToken> AuthenticateAsync(bool async, Token
             catch (ProbeRequestResponseException)
             {
                 // This was an expected response from the IMDS endpoint without the Metadata header set.
-                // Re-issue the request (CreateRequest will add the appropriate header).
+                // Fall-through to the code below to re-issue the request through MsalManagedIdentityClient.
+                continueIMDSRequestAfterProbe = true;
+            }
+
+            if (continueIMDSRequestAfterProbe)
+            {
+                try
+                {
 #pragma warning disable AZC0110 // DO NOT use await keyword in possibly synchronous scope.
-                var authResult = await _client.AcquireTokenForManagedIdentityAsync(context, cancellationToken).ConfigureAwait(false);
-                return authResult.ToAccessToken();
+                    var authResult = await _client.AcquireTokenForManagedIdentityAsync(context, cancellationToken).ConfigureAwait(false);
 #pragma warning restore AZC0110 // DO NOT use await keyword in possibly synchronous scope.
+                    return authResult.ToAccessToken();
+                }
+                catch (MsalServiceException e)
+                {
+                    if (e.Message.Contains("unavailable"))
+                    {
+                        throw new CredentialUnavailableException(IdentityUnavailableError, e);
+                    }
+                    throw new CredentialUnavailableException(UnknownError, e);
+                }
             }
+            throw new CredentialUnavailableException(UnknownError);
```

Comment:
Should these throw `AuthenticationFailedException` rather than `CredentialUnavailableException` for unknown errors to prevent credential chains from continuing? I know we don't use this distinction anymore in the development credentials but is it still needed for production credentials like ManagedIdentityCredential?

Reply:
The rationale for the fallback to CUE is that since this logic is implemented in `ImdsManagedIdentityProbeSource.cs` this would only happen in DAC, but not for pure `ManagedIdentityCredential`. So for more failure sensitive scenarios, the recommendation would be to use `ManagedIdentityCredential`